### PR TITLE
Adex : consent logs for adex RSP

### PR DESCRIPTION
**Pull request by kailash**
Consent logs for Adex:
- [x] Signing log with PKCS certificate 
- [ ] Logging consent flow (Apiserver --> optional consent log verticle)
- [ ] Pushing audit message to audit server

### CODE REVIEW

**Review by ThorodanBrom — APPROVED** [review]

**Comment by ThorodanBrom** [comment]
There is one more requirement as part of log creation - the Meity framework specifies that every log must have the base64 version of the consent artifact embedded in it. We cannot embed the entire consent artifact in the token since it's very large. Below is one way of solving this.

## Assumptions
- The RS proxy is registered on the auth server and has an admin user associated with it
- The admin user can get client credentials from the auth server and can use said credentials to get an admin token from the auth server that identifies the admin + the RS proxy
- The artifact ID is present in a valid PII data access token issued by the auth server 

## Flow
- When a user presents requests for PII data from the RS proxy, before the `DATA_REQUESTED` log is to be created, the RS proxy will get an admin token using client credentials from the auth server
- The RS proxy will then call a new API on the consent APD, where the user's PII data access token is in the request body and the admin token is used as authentication
- The consent APD validates the admin token + the PII data access token, extracts the artifact ID from the PII data access token, fetches from the DB and responds back the base64 version to the RS proxy
- The RS proxy can now use that response to create the `DATA_REQUESTED` log, as well as either `DATA_DENIED` or `DATA_SENT` log.

File: src/main/java/iudx/rs/proxy/apiserver/ApiServerVerticle.java [hunk 1]
```diff
@@ -110,6 +122,12 @@ public void start() throws Exception {
     attachCORSHandlers(router);
     attachDefaultResponses(router);
     router.route().handler(BodyHandler.create());
+    router.route().handler(new ConsentLogRequestHandler(true));
```
Comment (pranavrd) [inline src/main/java/iudx/rs/proxy/apiserver/ApiServerVerticle.java:125]: should the constructor parameter be configurable here?

Comment (kailash) [reply src/main/java/iudx/rs/proxy/apiserver/ApiServerVerticle.java:125]: Yes, It's configurable. you can read the `isAdexDeployed` property from the config and pass it as a constructor parameter.
Currently, In PR it is just kept as true, that needs to be changed before emerging PR.

File: src/main/java/iudx/rs/proxy/optional/consentlogs/dss/PayloadSigningManager.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+      List<DSSPrivateKeyEntry> keys = token.getKeys();
+      DSSPrivateKeyEntry privateKey = keys.get(0);
+
+      CAdESSignatureParameters signParameters = new CAdESSignatureParameters();
+      signParameters.setSignatureLevel(SignatureLevel.CAdES_BASELINE_B);
+      signParameters.setSignaturePackaging(SignaturePackaging.ENVELOPING);
+      signParameters.setDigestAlgorithm(DigestAlgorithm.SHA256);
+      signParameters.setSigningCertificate(privateKey.getCertificate());
+      signParameters.setCertificateChain(privateKey.getCertificateChain());
```
Comment (pranavrd) [inline src/main/java/iudx/rs/proxy/optional/consentlogs/dss/PayloadSigningManager.java:78]: no need for cert chain

File: src/main/java/iudx/rs/proxy/optional/consentlogs/ConsentLoggingVerticle.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package iudx.rs.proxy.optional.consentlogs;
+
+import io.vertx.core.AbstractVerticle;
+
+public class ConsentLoggingVerticle extends AbstractVerticle {
+  
+  @Override
+  public void start() throws Exception {
+
```
Comment (pranavrd) [inline src/main/java/iudx/rs/proxy/optional/consentlogs/ConsentLoggingVerticle.java:9]: PSM.init(config).getInstance();

File: src/main/java/iudx/rs/proxy/optional/consentlogs/dss/PayloadSigningManager.java (hunk 2 above)
Comment (ThorodanBrom) [reply src/main/java/iudx/rs/proxy/optional/consentlogs/dss/PayloadSigningManager.java:78]: Hello Pranav, if you're referring L78, `setCertificateChain` - it is better if that is present. After Mahidhar and I did some testing we found out that if the certificate chain is *not* added while signing, it does make the signed document smaller. But during validation, a lot of network calls are made to validate the document. The consent validator doesn't explicity do validation of data logs at this point, but if we ever do it, it's better to have the full chain. It's also good to have for posterity's sake.

**Comment by ananjaykumar2** [comment]
merged with #55

**Pull request closed without merging**